dart/dynamics/AssimpInputResourceAdaptor.h

```diff
@@ -0,0 +1,130 @@
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef DART_DYNAMICS_ASSIMPINPUTRESOURCEADAPTOR_H_
+#define DART_DYNAMICS_ASSIMPINPUTRESOURCEADAPTOR_H_
+#include <assimp/cfileio.h>
+#include <assimp/IOStream.hpp>
+#include <assimp/IOSystem.hpp>
+#include "dart/common/Resource.h"
+#include "dart/common/ResourceRetriever.h"
+
+namespace dart {
+namespace dynamics {
+
+class AssimpInputResourceRetrieverAdaptor : public Assimp::IOSystem
+{
+public:
+  AssimpInputResourceRetrieverAdaptor(
+    const common::ResourceRetrieverPtr& _resourceRetriever);
+  virtual ~AssimpInputResourceRetrieverAdaptor() = default;
+
+  /// \brief Tests for the existence of a file at the given path. 
+  /// 
+  /// \param pFile Path to the file
+  /// \return true if there is a file with this path, else false.
+  bool Exists(const char* pFile) const override;
+
+  ///  \brief Returns the system specific directory separator
+  ///
+  /// \return  System specific directory separator
+  char getOsSeparator() const override;
+
+  /// \brief Open a new file with a given path.
+  ///
+  /// When the access to the file is finished, call Close() to release
+  /// all associated resources (or the virtual dtor of the IOStream).
+  ///
+  /// \param pFile Path to the file
+  /// \param pMode Desired file I/O mode. Required are: "wb", "w", "wt",
+  ///        "rb", "r", "rt".
+  /// \return New IOStream interface allowing the lib to access
+  ///        the underlying file. 
+  Assimp::IOStream* Open(const char* pFile, const char* pMode = "rb") override;
+
+  /// \brief Closes the given file and releases all resources 
+  ///   associated with it.
+  /// \param pFile The file instance previously created by Open().
+  void Close(Assimp::IOStream* pFile) override;
+
+private:
+  common::ResourceRetrieverPtr mResourceRetriever;
+};
+
+class AssimpInputResourceAdaptor : public Assimp::IOStream
+{
+public:
+  AssimpInputResourceAdaptor(const common::ResourcePtr& _resource);
+  virtual ~AssimpInputResourceAdaptor() = default;
+
+  /// \brief Read from the file
+  ///
+  /// See fread() for more details
+  size_t Read(void* pvBuffer, size_t pSize, size_t pCount) override;
+
+  /// \brief Not implemented. This is a read-only stream.
+  size_t Write(const void* pvBuffer, size_t pSize, size_t pCount) override;
+
+  /// \brief Set the read/write cursor of the file
+  ///
+  /// Note that the offset is _negative_ for aiOrigin_END.
+  /// See fseek() for more details
+  aiReturn Seek(size_t pOffset, aiOrigin pOrigin) override;
+
+  /// \brief Get the current position of the read/write cursor
+  ///
+  /// See ftell() for more details
+  size_t Tell() const override;
+
+  /// \brief Returns filesize
+  ///
+  /// Returns the filesize.
+  size_t FileSize() const override;
+
+  /// \brief Not implemented. This is a read-only stream.
+  void Flush() override;
+
+private:
+  common::ResourcePtr mResource;
+};
+
+
+aiFileIO createFileIO(Assimp::IOSystem* adaptor);
+aiFile createFile(Assimp::IOStream* adaptor);
```

Comment:
This function can take *any* `Assimp::IOSystem` and wrap it in the C API. It is not specific to `AssimpInputResourceRetrieverAdaptor`. Therefore, I think it should remain a free function.